packages/nodejs/src/scope.ts

```diff
@@ -186,8 +212,12 @@ export class ScopeManager {
       if (oldScope === undefined) {
         this.removeSpanForUid(uid)
       } else {
-        this.#scopes.set(uid, oldScope)
-        this.#roots.set(uid, rootSpan)
+        if (rootSpan) {
+          this.setRoot(rootSpan)
+        }
+        if (oldScope) {
+          this.setActive(oldScope)
+        }
```

Comment:
Note that, since `.setRoot` and `.setActive` are used, and they only set spans if they're not closed, the old root and active spans will not be restored if they've been closed at this point in the code, meaning the span passed to `.withContext` will continue to be the active span after the `.withContext` callback has finished executing.

I don't think this is the desired behaviour. A closer-to-desired behaviour might be to also call `.removeSpanForUid` if either the root span or active span are closed.

Also note that, if the root span has been closed but the active span hasn't been closed, or the other way around, it will restore one but not the other, leading to a situation where the current root span may not be the root of the current active span. I don't think this is the desired behaviour either, I think that it should either restore both or remove both.

I think a possible implementation for the `finally` block that avoids these issues could be:

```typescript
finally {
  this.removeSpanForUid(uid)
  if (rootSpan && rootSpan.open && oldScope && oldScope.open) {
    this.setRoot(rootSpan)
    this.setActive(oldScope)
  }
}
```

Not blocking the PR on this, because there's other ways to end up in similarly inconsistent states using `.withContext`. It's kind of inevitable, at least with its current signature, as it takes an active span as an argument, but not a root span, and spans don't know their root spans.

But I think it would be best to avoid creating additional situations by which users of `.withContext` (or `tracer.withSpan` by proxy) can accidentally end up with an inconsistent state.

Reply:
I'm wondering if the root span is even something withContext should concern itself. Its behavior is to only overwrite the active span, why should it concern itself with the root span? If the root span gets closed during this time, it's not the withContext block responsibility to restore to a working state.